tools/codegen/gen.py

```diff
@@ -429,81 +376,43 @@ def go(f: NativeFunction) -> Optional[str]:
         assert len(f.func.arguments) > 0
         assert sum(a.name == 'self' for a in f.func.arguments) == 1
 
-        cpp_name = cpp.name(f.func)
-        cpp_returns_type = cpp.returns_type(f.func.returns)
-        signature_group = cpp.signature_group(f.func, method=True)
+        name = cpp.name(f.func)
+
+        sig_group = CppSignatureGroup.from_schema(f.func, method=True)
 
         if target is Target.DECLARATION:
-            if signature_group.gathered_signature is None:
-                # There's no TensorOptions. Just create the API without concern for TensorOptions.
-                return f"{cpp_returns_type} {cpp_name}({signature_group.signature.cpp_arguments_str(with_defaults=True)}) const;"
-            else:
-                # There's TensorOptions in the API. Create 2 APIs - one taking the TensorOptions object ("gathered_signature"),
-                # and one taking a scattered signature with ScalarType, Layout, Device separately ("signature").
-                # The gathered_signature already exists in several older PyTorch versions and had default arguments.
-                # For backward compatibility, we left it unchanged and added the scattered API on top of it.
-                # Note that the scattered API cannot have default arguments or calls will be ambigious.
-                return f"""
-{cpp_returns_type} {cpp_name}({signature_group.gathered_signature.cpp_arguments_str(with_defaults=True)}) const;
-{cpp_returns_type} {cpp_name}({signature_group.signature.cpp_arguments_str(with_defaults=False)}) const;
-"""
+            result = f"{sig_group.signature.decl()} const;\n"
```

Comment:
How did you manage to reduce the 2x2=4 branches down to 2? Last time we talked about this we came to the conclusion that even if you remove the proxies, you still need 4 branches. because you have 2 frontend signatures and 2 dispatcher signatures and you need to map from each to each.

Btw as a side note: Removing the proxies could have side implications because you now store the OperatorHandle singleton twice (once per function) instead of just once in the non-proxy function.

Reply:
> How did you manage to reduce the 2x2=4 branches down to 2? Last time we talked about this we came to the conclusion that even if you remove the proxies, you still need 4 branches. because you have 2 frontend signatures and 2 dispatcher signatures and you need to map from each to each.

I need 4 branches but it doesn't happen here, it happens at `cppargument_exprs`, see the "Scatter", "No op", "No op" and "Gather" cases. Because I organized things with argument packs, the decision between the branches can be done entirely in a data directed fashion.

> Btw as a side note: Removing the proxies could have side implications because you now store the OperatorHandle singleton twice (once per function) instead of just once in the non-proxy function.

Yes. I decided that besides a mild code size inflation this wouldn't really matter, and simplifying the codegen was more worth it.